feat: add `array/base/mskfilter-map` (#1705)
stdlib-js/stdlib: lib/node_modules/@stdlib/array/base/mskfilter-map/README.md

<!--

@license Apache-2.0

Copyright (c) 2024 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

> # mskfilter-map

Reviewer: ```suggestion
# mskfilterMap
```

> Apply a mask to a provided input array and return a new array after applying a mapping function.

<section class="usage">

## Usage

```javascript
var mskfilterMap = require( '@stdlib/array/base/mskfilter-map' );
```

#### mskfilterMap( x, mask, clbk\[, thisArg] )

> Returns a new array by applying a mask to a provided input array after applying mapping function.

Reviewer: ```suggestion
Applies a mask to a provided input array and returns a new array after applying a mapping function.
```

```javascript
var x = [ 1, 2, 3, 4 ];

function customMapping( value ) {
    return value * 2;
}

var y = mskfilterMap( x, [ 0, 1, 0, 1 ], customMapping );
// returns [ 4, 8 ]
```

The function supports the following parameters:

-   **x**: input array.
-   **mask**: mask array.
-   **clbk**: function to apply.
-   **thisArg**: applied function execution context (_optional_).

> To set the applied function's execution context, provide a `thisArg`.

Reviewer: You should provide an example here of providing a `thisArg`. A common example is counting the number of times the callback is invoked. You can copy the example above and then provide a context object which gets incremented after each invocation. For the example above, `clbk` should be called twice. See other base array packages.

Reviewer: @AuenKr I believe the README is still missing this example.

Reviewer: I see that you added it below in the `assign` API. I'd go ahead and move (and modify) that up here.

The function **always** returns a new "generic" array.

> #### mskfilterMap.assign( x, mask, out, stride, offset )

Reviewer: ```suggestion
#### mskfilterMap.assign( x, mask, out, stride, offset, clbk\[, thisArg] )
```

> Applies a mask & mapping function to a provided input array and assigns unmasked values to elements in a provided output array.

Reviewer: ```suggestion
Applies a mask and mapping function to a provided input array and assigns results to elements in a provided output array.
```

```javascript
var x = [ 1, 2, 3, 4 ];
var mask = [ 0, 1, 0, 1 ];

var out = [ 0, 0, 0, 0 ];

function clbk( val ) {
    return val * 2;
}

var arr = mskfilterMap.assign( x, mask, out, -2, out.length-1, clbk );
// returns [ 0, 8, 0, 4 ]

var bool = ( arr === out );
// returns true
```

The function supports the following parameters:

-   **x**: input array.
-   **mask**: mask array.
-   **out**: output array.
-   **stride**: output array stride.
-   **offset**: output array offset.
-   **clbk**: function to apply.
-   **thisArg**: applied function execution context (_optional_).

</section>

<!-- /.usage -->

<section class="notes">

## Notes

-   If a `mask` array element is truthy, the corresponding element in `x` is **included** in the output array; otherwise, the corresponding element in `x` is "masked" and thus **excluded** from the output array.

</section>

<!-- /.notes -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var zeroTo = require( '@stdlib/array/base/zero-to' );

> var bernoulli = require( '@stdlib/random/array/bernoulli' );

Reviewer: ```suggestion
var bernoulli = require( '@stdlib/random/array/bernoulli' );
var abs2 = require( '@stdlib/math/base/special/abs2' );
```

Use the stdlib package.

var mskfilterMap = require( '@stdlib/array/base/mskfilter-map' );

// Generate a linearly spaced array:
var x = zeroTo( 20 );
console.log( x );

// Generate a random mask:
var mask = bernoulli( x.length, 0.5, {
    'dtype': 'generic'
});
console.log( mask );

> // Mapping functions
> function square( val ) {
>     return val * val;
> }
> 

Reviewer: ```suggestion
```

// Filter an array using the mask:

> var y = mskfilterMap( x, mask, square );

Reviewer: ```suggestion
var y = mskfilterMap( x, mask, abs2 );
```

<!--

@license Apache-2.0

Copyright (c) 2024 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# mskfilterMap

> Apply a mask to a provided input array and return a new array after applying a mapping function.

<section class="usage">

## Usage

```javascript
var mskfilterMap = require( '@stdlib/array/base/mskfilter-map' );
```

#### mskfilterMap( x, mask, clbk\[, thisArg] )

Applies a mask to a provided input array and returns a new array after applying a mapping function.

```javascript
function clbk( value ) {
    return value * 2;
}

> var x = [ 1, 2, 3, 4 ];
> 
> var y = mskfilterMap( x, [ 0, 1, 0, 1 ], clbk );

Reviewer: ```suggestion
var x = [ 1, 2, 3, 4 ];
var m = [ 0, 1, 0, 1 ];

var y = mskfilterMap( x, m, clbk );
```

The idea behind these changes is to make it visually more obvious what values are masked and to minimize eye movement when interpreting output results.

> var x = [ 1, 2, 3, 4 ];
> 
> function customMapping( value ) {
>     return value * 2;
> }
> 
> var y = mskfilterMap( x, [ 0, 1, 0, 1 ], customMapping );

Reviewer: ```suggestion
function clbk( value ) {
    return value * 2;
}

var x = [ 1, 2, 3, 4 ];

var y = mskfilterMap( x, [ 0, 1, 0, 1 ], clbk );
```

var y = mskfilterMap( x, m, clbk );
// returns [ 4, 8 ]
```

The function supports the following parameters:

-   **x**: input array.
-   **mask**: mask array.
-   **clbk**: function to apply.
-   **thisArg**: applied function execution context (_optional_).

To set the applied function's execution context, provide a `thisArg`.

The function **always** returns a new "generic" array.

#### mskfilterMap.assign( x, mask, out, stride, offset, clbk\[, thisArg] )

Applies a mask and mapping function to a provided input array and assigns results to elements in a provided output array.

```javascript
function clbk( value ) {
    return value * 2;
}

var x = [ 1, 2, 3, 4 ];
var m = [ 0, 1, 0, 1 ];

var out = [ 0, 0, 0, 0 ];
var arr = mskfilterMap.assign( x, m, out, -2, out.length-1, clbk );
// returns [ 0, 8, 0, 4 ]

var bool = ( arr === out );
// returns true
```

The function supports the following parameters:

-   **x**: input array.
-   **mask**: mask array.
-   **out**: output array.
-   **stride**: output array stride.
-   **offset**: output array offset.
-   **clbk**: function to apply.
-   **thisArg**: applied function execution context (_optional_).

> To set the applied function's execution context, provide a `thisArg`.

Reviewer: You can remove this line once you've moved the example below up to the previous API.

<!-- eslint-disable no-invalid-this -->

```javascript
function clbk( x ) {
    this.count += 1;
    return x;
}

> var x = [ 1.0, 2.0, 3.0, 4.0 ];
> var mask = [ 1, 1, 0, 1 ];

Reviewer: ```suggestion
var x = [ 1, 2, 3, 4 ];
var m = [ 1, 1, 0, 1 ];
```

Consistency with examples above.

var ctx = {
    'count': 0
};

> var y = mskfilterMap( x, mask, clbk, ctx );

Reviewer: ```suggestion
var y = mskfilterMap( x, m, clbk, ctx );
```

> // returns [ 1.0, 2.0, 4.0 ]

Reviewer: ```suggestion
// returns [ 1, 2, 4 ]
```

limitations under the License.

-->

# mskfilterMap

> Apply a mask to a provided input array and return a new array after applying a mapping function.

<section class="usage">

## Usage

```javascript
var mskfilterMap = require( '@stdlib/array/base/mskfilter-map' );
```

#### mskfilterMap( x, mask, clbk\[, thisArg] )

Applies a mask to a provided input array and returns a new array after applying a mapping function.

```javascript
function clbk( value ) {
    return value * 2;
}

var x = [ 1, 2, 3, 4 ];
var m = [ 0, 1, 0, 1 ];

var y = mskfilterMap( x, m, clbk );
// returns [ 4, 8 ]
```

The function supports the following parameters:

-   **x**: input array.
-   **mask**: mask array.
-   **clbk**: function to apply.
-   **thisArg**: applied function execution context (_optional_).

To set the applied function's execution context, provide a `thisArg`.

The function **always** returns a new "generic" array.

#### mskfilterMap.assign( x, mask, out, stride, offset, clbk\[, thisArg] )

Applies a mask and mapping function to a provided input array and assigns results to elements in a provided output array.

```javascript

> var x = [ 1, 2, 3, 4 ];
> var mask = [ 0, 1, 0, 1 ];
> 
> var out = [ 0, 0, 0, 0 ];
> 
> function clbk( val ) {
>     return val * 2;
> }
> 
> var arr = mskfilterMap.assign( x, mask, out, -2, out.length-1, clbk );

Reviewer: ```suggestion
function clbk( value ) {
    return value * 2;
}

var x = [ 1, 2, 3, 4 ];
var m = [ 0, 1, 0, 1 ];

var out = [ 0, 0, 0, 0 ];
var arr = mskfilterMap.assign( x, m, out, -2, out.length-1, clbk );
```

Same general comment as above.
